### PR TITLE
Adding image cleanup for clouds

### PR DESCRIPTION
**Pull request by akhil-jha**
Links HMS-2043
- [x] Azure (depends on https://github.com/RedHatQE/wrapanapi/pull/465 )
- [x] AWS

```

(cloudwash) ➜  cloudwash git:(support_image_removal) ✗ swach -d aws --all

<<<<<<< Running the cleanup script in DRY RUN mode >>>>>>>
The AWS providers settings are initialized and validated !

Resources from the region: us-east-1

=========== DRY SUMMARY ============

IMAGES:
	Deletable: ['ami-0ba22849118c7101a', 'ami-03d1c1aee99de0c58']

====================================

(cloudwash) ➜  cloudwash git:(support_image_removal) ✗ swach -d azure --all

<<<<<<< Running the cleanup script in DRY RUN mode >>>>>>>
The AZURE providers settings are initialized and validated !

Resources from the region and resource group: eastus/redhat-deployed

=========== DRY SUMMARY ============

DISCs:
	Deletable: ['composer-api-qwd_disk1_9556a6aca0324ac5adf7e4a700f1c09b']
NICs:
	Deletable: ['composer-api-qwd96']
IMAGES:
	Deletable: ['composer-api-985124e5-b8a9-4969-98ec-400b6af7c924']

====================================

(cloudwash) ➜  cloudwash git:(support_image_removal) ✗ swach azure --all 

<<<<<<< Running the cleanup script in ACTION RUN mode >>>>>>>
The AZURE providers settings are initialized and validated !

Resources from the region and resource group: eastus/redhat-deployed
Stopped VMs: 
[]
Removed VMs: 
[]
Skipped VMs: 
[]
Removed NICs: 
['composer-api-qwd96']
Removed Discs: 
['composer-api-qwd_disk1_9556a6aca0324ac5adf7e4a700f1c09b']
Removed PIPs: 
['composer-api-qwd-ip']
Removed Images: 
['composer-api-985124e5-b8a9-4969-98ec-400b6af7c924']
```

Can't run on aws as we need those images :P

### CODE REVIEW

**Comment by akhil-jha** [comment]
```
swach azure --all

<<<<<<< Running the cleanup script in ACTION RUN mode >>>>>>>
The AZURE providers settings are initialized and validated !

Resources from the region and resource group: eastus/redhat-deployed
Stopped VMs: 
[]
Removed VMs: 
[]
Skipped VMs: 
[]
Removed Images: 
['composer-api-0b2a4123-68d5-4be7-9b71-e88401f90b4c', 'composer-api-48b46ce3-0e8d-42f4-ae35-f3e8a84167c3', 'composer-api-4d3c79dc-bebe-4688-971d-1baa9498c92f', 'composer-api-5181f605-0e5f-4a4e-9a9d-3f5b5de8ea7b', 'composer-api-596f8347-ea4e-447a-87b6-deff11ad6e7c', 'composer-api-5cb99b06-f685-46ba-9270-5c6cdc70401a', 'composer-api-6a46dda6-ef02-43f5-9973-1dd77a9e1578', 'composer-api-717af3b9-4c60-4f1e-be22-9af703fff0fe', 'composer-api-83731b17-41a0-446d-b56b-e56a0ddaeafd', 'composer-api-9947ab7a-5ef9-4041-bb64-316d156436cf', 'composer-api-ab2aa738-d362-4ebb-8388-9ab5c62d9a69', 'composer-api-c544f5c2-7ab0-428e-a085-fad79c1c559e', 'composer-api-d821a29f-c06d-442e-ad1c-df5b36464bff', 'composer-api-d942293b-1a80-4516-9f43-b08892a4acd7', 'composer-api-daf2fb61-8621-49ca-867e-1e75af0c2e41', 'composer-api-e2ee5537-a969-4d33-bf2b-7b34a4bc8501', 'composer-api-e471154d-50bf-4899-8c86-dbb5f3d522a6', 'composer-api-e4fadca4-41ab-4323-ba38-d8ff8421115a', 'composer-api-fc0d890a-896c-489b-892a-6ba136d940d3', 'composer-api-fed68239-5d7e-43f8-ab89-003fedd25e81']
(cloudwash) ➜  cloudwash git:(support_image_removal) ✗ swach -d aws --all

<<<<<<< Running the cleanup script in DRY RUN mode >>>>>>>
The AWS providers settings are initialized and validated !

Resources from the region: us-east-1

=========== DRY SUMMARY ============

IMAGES:
	Deletable: ['ami-03d1c1aee99de0c58', 'ami-0ba22849118c7101a']

====================================


```

File: cloudwash/providers/aws.py [hunk 1]
```diff
@@ -53,6 +53,27 @@ def dry_discs():
...
+                    remove_images = [
+                        image
+                        for image in free_images
+                        if image not in settings.aws.exceptions.images
+                    ]
+                    if settings.aws.criteria.image.delete_pattern:
+                        remove_images = [
+                            image
+                            for image in remove_images
+                            if not image.startswith(settings.aws.criteria.image.delete_pattern)
```
Comment (mshriver) [inline cloudwash/providers/aws.py:72]: This `not` looks backwards.

If I have a list of free images that aren't in the exception list, and a `delete_pattern` from config that dictates what image names to delete, this gives me a list of names to keep, not names to remove.

```
# list of free images not excluded
In [1]: free_list = ['deleteme-one', 'image-two', 'image-three', 'deleteme-four']

# assuming delete_pattern is 'deleteme'
In [2]: remove_images = [i for i in free_list if not i.startswith('deleteme')]

# I end up with the images that should keep, not that should be deleted?
In [3]: remove_images
Out[3]: ['image-two', 'image-three']

```

Comment (mshriver) [reply cloudwash/providers/aws.py:72]: Maybe this comes from an intent to have the default behavior pick all images for removal if the `delete_pattern` is empty?

Either way, it needs to be adjusted so `delete_pattern` is applied correctly when present - I would recommend actually using `re` here and actually accept a pattern string. Using just `startswith` will immediately limit the use and dictate that any teams wanting to use this tool have to prefix image names.

`re.compile` can be used to create the regex pattern object (and to parse the pattern from config), and then you can use  `Pattern.search` function in the list comprehension to indicate whether the given pattern is anywhere in the string. `match` can be used too, but will be more explicit and will require the caller to specify `.*` in their pattern to match anywhere in the string.  I think either function is fine here, but the docblock/readme/config comment should explain which is in use.

Comment (jyejare) [reply cloudwash/providers/aws.py:72]: Also agree with Mike here, with `not` we are doing opposite here! Please remove `not` from the condition!

**Review by jyejare — CHANGES_REQUESTED** [review]
ACK pending comments!

File: cloudwash/providers/azure.py [hunk 2]
```diff
@@ -97,6 +97,30 @@ def dry_resources(hours_old=None):
                     )
                     return dry_data["RESOURCES"]["delete"]
 
+                def dry_images():
+                    remove_images = []
+                    if settings.azure.criteria.image.unassigned:
+                        images_list = azure_client.list_compute_images_by_resource_group(
```
[the remark this answers is not in the document]
Comment (akhil-jha) [reply cloudwash/providers/azure.py:103]: Continuing from wrapanapi comment, I don't think there is a need for creating two methods for it. free_images flag can be used to change the desired output

File: cloudwash/providers/aws.py (hunk 1 above)
Comment (akhil-jha) [reply cloudwash/providers/aws.py]: Removed not.
As for delete_pattern, I am moving away from it and keep it what cloudwash already has. Deleting images starting with a keyword mentioned in conf.

File: cloudwash/providers/azure.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jyejare) [reply cloudwash/providers/azure.py:103]: Not blocking your PR on it !

**Pull request merged**